pysmiles/cheminfo.py

```diff
@@ -0,0 +1,182 @@
+"""
+Some simple cheminformatics tools.
+"""
+from collections import defaultdict
+import numpy as np
+import networkx as nx
+from .smiles_helper import PTE, remove_explicit_hydrogens
+
+def tanimoto_similarity(binarr1, binarr2):
+    """
+    Compute Tanimoto similarity between two binary arrays.
+
+    Parameters
+    ----------
+    binarr1: np.array
+    binarr2: np.array
+
+    Returns
+    -------
+    float
+        the tanimoto similarity
+    """
+    intersection = np.sum(binarr1 & binarr2)
+    union = np.sum(binarr1) + np.sum(binarr2) - intersection
+    return intersection / union if union != 0 else 0.0
+
+def features_to_binary(feature_list, size=1024):
+    """
+    Converte a list of features to a binary array of
+    given `size`.
+
+    Parameters
+    ----------
+    feature_list: abc.iteratable[int]
+        a iteratable with integer numbers
+    size: int
+        size of the binary array; default=1024
+
+    Returns
+    -------
+    np.array
+    """
+    fp_array = np.zeros(size, dtype=int)
+    nonzero = np.array(feature_list, dtype=int) % size
+    fp_array[nonzero] = 1
+    return fp_array
+
+def _aggregate_features(graph, cycles=4, label='fp_invariant'):
+    """
+    Given atomic invariants in a molecule stored under the `label`
+    keyword, iteratively aggregate those features for a given number
+    of `cycles`. It returns a feature list, which is composed of the
+    hashes of the aggregated features.
+
+    Parameters
+    ----------
+    graph: networkx.Graph
+    cycles: int
+        number of cycles to run; default is 4
+    label: abc.hashable
+        the keyword under which the inital invariants are stored
+
+    Returns
+    -------
+    list[int]
+        list of the hashes of unique features
+    """
+    feature_list = list(nx.get_node_attributes(graph, label).values())
+    substructures = set(nx.get_node_attributes(graph, 'subgraph').values())
+    for idx in range(0, cycles):
+        for node in graph.nodes:
+            subgraph = set(graph.nodes[node]['subgraph'])
+            feat_list_iter = [(idx, graph.nodes[node][label])]
+            for neigh in graph.neighbors(node):
+                order = graph.edges[(node, neigh)].get('order', 1)
+                feat_list_iter.append((order, graph.nodes[neigh][label]))
+                subgraph = subgraph | graph.nodes[neigh]['subgraph']
+            graph.nodes[node]['subgraph'] = subgraph
+            feat_list_iter = sorted(feat_list_iter)
+            feat_tuple_iter = tuple(feat for sub_feat in feat_list_iter for feat in sub_feat)
+            new_hash = hash(feat_tuple_iter)
+            graph.nodes[node]['fp_invariant'] = new_hash
+            if frozenset(subgraph) not in substructures:
+                feature_list.append(new_hash)
+                substructures.add(frozenset(subgraph))
```

Comment:
You don't use the substructures, do you?
Instead of just doing `new_hash = hash(feat_tuple_iter)`, first try to lookup the hash for this fragment in your substructures. Substructures should then be a dict of frozenset(subgraph) -> hash

Reply:
You do because the hashes will be different depending on which atom you start even if they generate the same substructure

Reply:
Right. I missed that you append to feature_list here, only if your subgraph is new